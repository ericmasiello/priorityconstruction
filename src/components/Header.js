import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import GatsbyImage from './GatsbyImage';

const BackgroundImage = GatsbyImage.extend`
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  opacity: 0.3;
`;

const HeaderContainer = styled.div`
  margin: 0 auto;
  max-width: 960px;
  padding: 1.45rem 1.0875rem;
  position: relative;
  z-index: 2;
  display: flex;
  justify-content: space-between;
`;

export class Header extends Component {
  static displayName = 'Header';

  static propTypes = {
    tag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
    children: PropTypes.node,
    isFullHeight: PropTypes.bool,
    bgImageSizes: PropTypes.shape({
      src: PropTypes.string.isRequired,
      base64: PropTypes.string,
      aspectRatio: PropTypes.number,
      sizes: PropTypes.string,
      srcSet: PropTypes.string,
    }).isRequired,
  };

  static defaultProps = {
    tag: 'header',
  };

  componentDidUpdate = (prevProps) => {
    const { isFullHeight } = this.props;
    if (isFullHeight !== prevProps.isFullHeight) {
      if (isFullHeight) {
        this.elm.animate([{ height: '20vh' }, { height: '70vh' }], {
          duration: 300,
          fill: 'forwards',
          easing: 'cubic-bezier(0.86, 0, 0.07, 1)',
          iterations: 1,
        });
      } else {
        this.elm.animate([{ height: '70vh' }, { height: '20vh' }], {
          duration: 300,
          fill: 'forwards',
          easing: 'cubic-bezier(0.86, 0, 0.07, 1)',
          iterations: 1,
        });
      }
    }
  };

  render() {
    const {
      bgImageSizes,
      isFullHeight,
      children,
      tag: Tag,
      ...rest
    } = this.props;
    return (
      <Tag
        ref={(elm) => {
          // eslint-disable-next-line react/no-find-dom-node
          this.elm = ReactDOM.findDOMNode(elm);
        }}
        {...rest}
      >
        <HeaderContainer>
          {children}
        </HeaderContainer>
        <BackgroundImage
          style={{
            position: 'absolute',
          }}
          sizes={bgImageSizes}
        />
      </Tag>
    );
  }
}

export default styled(Header)`
  background: #524763;
  margin-bottom: 1.45rem;
  overflow: hidden;
  position: relative;
  height: ${({ isFullHeight }) => (isFullHeight ? '70vh' : '20vh')};
  h1 {
    img {
      height: 80px;
    }
  }
`

