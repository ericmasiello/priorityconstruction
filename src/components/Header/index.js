import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import Link from 'gatsby-link';
import styled from 'styled-components';
import Img from 'gatsby-image';

const HeaderWrapper = styled.div`
  background: #524763;
  margin-bottom: 1.45rem;
  overflow: hidden;
  position: relative;
  height: ${({ isHome }) => (isHome ? '70vh' : '20vh')};
  h1 {
    img {
      height: 80px;
    }
  }
`;

const HeaderContainer = styled.div`
  margin: 0 auto;
  max-width: 960px;
  padding: 1.45rem 1.0875rem;
  position: relative;
  z-index: 2;
  display: flex;
  justify-content: space-between;
`;

const Logo = styled.h1`
  ${props => `background-image: url('${props.image}');`}
  width: 100%;
  height: 100%;
  background-size: contain;
  text-indent: -9999px;
  overflow: hidden;
  margin: 0;
`;

const LogoLink = styled(Link)`
  display: inline-block;
  height: 100px;
  width: 440px;
`;

const MainNav = styled.nav`
  ul {
    list-style: none;
    display: flex;
    li {
      margin-left: 10px;
      font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen, Ubuntu, Cantarell,
        Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
      a {
        text-decoration: none;
        color: #fff;
        &:hover {
          border-bottom: 3px solid #524763;
        }
      }
    }
  }
`;

export default class Header extends Component {
  static propTypes = {
    title: PropTypes.string,
    location: PropTypes.shape({
      pathname: PropTypes.string,
    }).isRequired,
    data: PropTypes.shape({
      background: PropTypes.shape({
        sizes: PropTypes.shape({}),
      }),
      logo: PropTypes.shape({
        sizes: PropTypes.shape({}),
      }),
    }).isRequired,
  };
  componentDidUpdate = (prevProps) => {
    const { location } = this.props;
    if (location.pathname !== prevProps.location.pathname) {
      if (this.props.location.pathname === '/') {
        this.wrapper.animate([{ height: '20vh' }, { height: '70vh' }], {
          duration: 300,
          fill: 'forwards',
          easing: 'cubic-bezier(0.86, 0, 0.07, 1)',
          iterations: 1,
        });
      } else {
        this.wrapper.animate([{ height: '70vh' }, { height: '20vh' }], {
          duration: 300,
          fill: 'forwards',
          easing: 'cubic-bezier(0.86, 0, 0.07, 1)',
          iterations: 1,
        });
      }
    }
  };

  render() {
    const { title, data, location } = this.props;
    return (
      <HeaderWrapper
        isHome={location.pathname === '/'}
        ref={(wrapper) => {
          /* eslint-disable react/no-find-dom-node */
          this.wrapper = ReactDOM.findDOMNode(wrapper);
          /* eslint-enable react/no-find-dom-node */
        }}
      >
        <HeaderContainer>
          <LogoLink to="/">
            <Logo image={data.logo.sizes.src}>
              {title}
            </Logo>
          </LogoLink>
          <MainNav>
            <ul>
              <li>
                <Link to="/">Home</Link>
              </li>
              <li>
                <Link to="/about">About</Link>
              </li>
            </ul>
          </MainNav>
        </HeaderContainer>
        <Img
          style={{
            position: 'absolute',
            left: 0,
            top: 0,
            width: '100%',
            height: '100%',
            opacity: 0.3,
          }}
          sizes={data.background.sizes}
        />
      </HeaderWrapper>
    );
  }
}
