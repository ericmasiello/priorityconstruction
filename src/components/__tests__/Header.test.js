import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import 'jest-styled-components';
import Header, { Header as Raw } from '../Header';

const props = {
  bgImageSizes: {
    src: '/foo/bar.jpg',
  },
};

it('should render', () => {
  const component = renderer.create(<Header {...props} />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render children', () => {
  const wrapper = mount(<Raw {...props}>Test</Raw>);

  expect(wrapper.text()).toContain('Test');
});

it('should render additional props', () => {
  const wrapper = shallow(<Raw {...props} data-test="my-attribute" />);

  expect(wrapper.find('[data-test="my-attribute"]')).toHaveLength(1);
});

it('should display additional classes', () => {
  const wrapper = shallow(<Raw {...props} className="my-class" />);

  expect(wrapper.hasClass('my-class')).toBe(true);
});

it('should render as a <header> by default', () => {
  const wrapper = shallow(<Raw {...props} />);

  expect(wrapper.type()).toBe('header');
});

it('should render with a custom tag', () => {
  const wrapper = shallow(<Raw {...props} tag="span" />);

  expect(wrapper.type()).toBe('span');
});
