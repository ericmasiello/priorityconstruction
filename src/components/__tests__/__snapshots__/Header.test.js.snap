// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
.c2 {
  max-width: 100%;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  opacity: 0.3;
}

.c1 {
  margin: 0 auto;
  max-width: 960px;
  padding: 1.45rem 1.0875rem;
  position: relative;
  z-index: 2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  background: #524763;
  margin-bottom: 1.45rem;
  overflow: hidden;
  position: relative;
  height: 20vh;
}

.c0 h1 img {
  height: 80px;
}

<header
  className="c0"
>
  <div
    className="c1"
  />
  <div
    className=" gatsby-image-outer-wrapper"
    style={
      Object {
        "position": "initial",
      }
    }
  >
    <div
      className="c2 gatsby-image-wrapper"
      style={
        Object {
          "overflow": "hidden",
          "position": "absolute",
        }
      }
    >
      <div
        style={
          Object {
            "paddingBottom": "NaN%",
            "width": "100%",
          }
        }
      />
      <img
        alt=""
        onLoad={[Function]}
        src="/foo/bar.jpg"
        style={
          Object {
            "height": "100%",
            "left": 0,
            "objectFit": "cover",
            "objectPosition": "center",
            "opacity": 1,
            "position": "absolute",
            "top": 0,
            "transition": "opacity 0.5s",
            "width": "100%",
          }
        }
      />
      <noscript
        dangerouslySetInnerHTML={
          Object {
            "__html": "<img src=\\"/foo/bar.jpg\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/>",
          }
        }
      />
    </div>
  </div>
</header>
`;
